### PR TITLE
BUG/API: .merge() and .join() on category dtype columns will now preserve category dtype

### LINKED ISSUE #10409: BUG: merge with categoricals does not preserve categories dtype
xref #14351

None of the following merge operations retain the `category` types. Is this expected? How can I keep them?
#### Merging on a `category` type:

Consider the following:

```
A = pd.DataFrame({'X': np.random.choice(['foo', 'bar'],size=(10,)), 
                  'Y': np.random.choice(['one', 'two', 'three'], size=(10,))})
A['X'] = A['X'].astype('category')

B = pd.DataFrame({'X': np.random.choice(['foo', 'bar'],size=(10,)), 
                  'Z': np.random.choice(['jjj', 'kkk', 'sss'], size=(10,))})
B['X'] = B['X'].astype('category')
```

if I do the merge, we end up with:

```
> pd.merge(A, B, on='X').dtypes 
X    object
Y    object
Z    object
dtype: object
```
#### Merging on a `non-category` type:

```
A = pd.DataFrame({'X': np.random.choice(['foo', 'bar'],size=(10,)), 
                  'Y': np.random.choice(['one', 'two', 'three'], size=(10,))})
A['Y'] = A['Y'].astype('category')

B = pd.DataFrame({'X': np.random.choice(['foo', 'bar'],size=(10,)), 
                  'Z': np.random.choice(['jjj', 'kkk', 'sss'], size=(10,))})
B['Z'] = B['Z'].astype('category')
```

if I do the merge, we end up with:

```
pd.merge(A, B, on='X').dtypes
X    object
Y    object
Z    object
dtype: object
```

### PR DESCRIPTION
**Pull request by jreback**
closes #10409

```
    before     after       ratio
  [6d2293f7] [c352573f]
-  410.11ms   260.46ms      0.64  join_merge.MergeCategoricals.time_merge_cat

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

### CODE REVIEW

**Comment by jreback** [comment]
cc @amelio-vazquez-reina  
cc @psychemedia
cc @watercrossing 

as you have posted issues about this in the past.

**Comment by jreback** [comment]
@jorisvandenbossche if you'd have a look (mainly at the tests).

**Comment by chris-b1** [comment]
Do you want to add an asv for this?

Probably a separate issue, but should we be warning about the implicit conversion to `object` when the categories don't match?

**Comment by jreback** [comment]
@chris-b1 yes I should add an asv. that's easy.

> Probably a separate issue, but should we be warning about the implicit conversion to object when the categories don't match?

hmm, we don't do this generally on merges, though maybe we should. Can you open another issue (with some examples), even things like string/int merging should be warn/raised I think. Maybe need a ``errors='warn|raise'`` generally in merge.

**Comment by jreback** [comment]
@chris-b1 added benchmark.

File: pandas/tools/tests/test_merge.py [hunk 1]
```diff
@@ -1368,3 +1369,118 @@ def f():
...
+
+        result = result.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           np.dtype('O'),
+                           np.dtype('O')],
+                          index=['X', 'Y', 'Z'])
+        assert_series_equal(result, expected)
+
+        # swap the categories and ordered on one
+        # but should still work (end return categorical)
```
Comment (jorisvandenbossche) [inline pandas/tools/tests/test_merge.py]: I don't think this one should work? 
For concat it returns object (as the categories are not identical)

File: pandas/tools/tests/test_merge.py [hunk 2]
```diff
@@ -1368,3 +1369,118 @@ def f():
...
+                              pd.Index(['bar', 'foo']))
+
+        assert_frame_equal(result, merged)
+
+        result = result.dtypes.sort_index()
+        expected = Series([CategoricalDtype(),
+                           np.dtype('O'),
+                           np.dtype('O')],
+                          index=['X', 'Y', 'Z'])
+        assert_series_equal(result, expected)
```
Comment (jorisvandenbossche) [inline pandas/tools/tests/test_merge.py]: Should there be a test here added with actual different categories that results in object type of the merge column X

**Review by jorisvandenbossche — COMMENTED** [review]
Took a look at the tests, and added few questions. 
I have to say that the tests are a bit hard to follow due to all the asigning and manipulation of the original left/right frames

**Comment by jreback** [comment]
so after looking at this there are some cases that we ought to consider.

assume we are joining with ``left`` and ``right`` with a single ``on`` (it can be the index or not, but ultimately these come down to a left and right on a column (single) for now.

- ``how='left'``: will be the dtype of the left, regardless of right
- ``how='right'``: will be the dtype of the right, regardless of left
- ``how='inner'``: if left and right are match on dtype (IOW if categorical then categories & ordered must match), OR right is ``object``, then can return ``categorical`` of left flavor, otherwise must be ``object``
- ``how='outer'``: I think same as inner, these must match exactly (or be ``object`` that has categories that are fully contained within ``left``).

So I think these follow our rules exactly from [here, on the bottom](http://pandas-docs.github.io/pandas-docs-travis/categorical.html#concatenation)

**Comment by jorisvandenbossche** [comment]
Ah, yes, I didn't consider the different join operations. 

> how='inner': if left and right are match on dtype (IOW if categorical then categories & ordered must match), OR right is object, then can return categorical of left flavor, otherwise must be object

For inner and outer, I would follow the same rules as `concat`: this means only retaining category dtype if categories and orderedness match exactly, so for both left and right frame. 
(the "be object that has categories that are fully contained within the other" rule is something we removed from `concat`, so I wouldn't introduce that again?)

**Comment by jreback** [comment]
@jorisvandenbossche ok revised. If you'd play with this when you have a chance.

**Comment by jreback** [comment]
@jorisvandenbossche if you have a chance

**Comment by jreback** [comment]
@jorisvandenbossche if you'd have a look. going to merge soon.

File: pandas/tests/types/test_common.py [hunk 3]
```diff
@@ -39,16 +39,33 @@ def test_period_dtype(self):
...
+
+    for dtype in [np.float32]:
+        assert not is_dtype_equal(dtypes['f'], dtype)
+
+    # strict w.r.t. PeriodDtype
+    assert not is_dtype_equal(dtypes['p'], PeriodDtype('2D'))
+
+    # strict w.r.t. datetime64
+    assert not is_dtype_equal(dtypes['dt_tz'],
+                              pandas_dtype('datetime64[ns, CET]'))
```
Comment (jorisvandenbossche) [inline pandas/tests/types/test_common.py]: Is there a reason there are no categorical dtype in this test?

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

I would just mention this somewhere in the merging and/or categorical docs

File: pandas/tests/types/test_common.py (hunk 3 above)
Comment (jreback) [inline pandas/tests/types/test_common.py]: no will add

**Comment by has2k1** [comment]
Some feedback. I had this -- 1.5 year old -- function that I have used to compensate for the bug

```python
# NOTE: This is a temporary fix due to bug
# https://github.com/pydata/pandas/issues/10409
# Remove when that bug is fixed
import pandas.api.types as pdtypes

def preserve_categories(ref, other):
    for col in ref.columns & other.columns:
        if pdtypes.is_categorical_dtype(ref[col]):
            other[col] = other[col].astype(
                'category', categories=ref[col].cat.categories)
```

I have removed it and my tests pass.
